### ISSUE TITLE
Arithmetic expression being dropped

### ISSUE THREAD

**Issue by InusualZ** [matt-kempster/m2c#243]
*Sorry about the tittle, I don't know enough about how the decompiler works so I went with what I understand is happening*

m2c output
![image](https://user-images.githubusercontent.com/2200611/185799612-8431ca59-865c-4872-9a90-7a44ddcf8d70.png)

Ghidra Output
![image](https://user-images.githubusercontent.com/2200611/185799533-a31e94c0-8bc5-451b-ac66-c4f87a32624a.png)

arguments
```
python m2c.py -t ppc-mwcc-c --context context.c input.s
```

[input.s](https://github.com/matt-kempster/m2c/files/9389477/input.txt)
[context.c](https://github.com/matt-kempster/m2c/files/9389478/context.txt)

**Comment by simonlindholm** [comment]
Looks like we don't recognize that r5 is an argument to sprintf, so we optimize it out. Knowing which registers to pass as arguments to functions on ppc is difficult in general, unfortunately, though we should be able to add a heuristic that captures this particular case based on the deduction that the register is not used in any other way.

**Comment by InusualZ** [comment]
Ohh, I see.

I changed the function definition to:
```
void sprintf(char* __s, char* __format, unsigned int d); // @8045decc
```

from: 
```
void sprintf(char* __s, char* __format, ...); // @8045deccZ
```

And now it "work"
![image](https://user-images.githubusercontent.com/2200611/185800660-a32b4f72-3d3e-47c3-bfea-dac12496863a.png)

Maybe you can see also check if there is a function with variadic arguments

**Comment by simonlindholm** [comment]
And have variadic functions consume all the registers they can? Yeah, that's not unreasonable. Will lead to false positives but that may be okay given variadic functions are a rare case.

**Comment by InusualZ** [comment]
>  consume all the registers they can

I don't know if possible, but you can exclude most registers that haven't being touched since the last call to a function (before the call to the variadic)

**Comment by simonlindholm** [comment]
Yeah, we already take into account function calls clobbering registers and such. But argument registers are very often used as temps, and called unknown functions are treated as writing to r3/r4/f1.

**Comment by simonlindholm** [comment]
Fixed in 008adfcf9afeb0ea6df2cbee3d2351c832cd0bef (for variadic functions, lmk if this generates too many false positives and I can back it out) and 6b6e1390bf30013247343623194964ff13d0c45b (for functions not in context, based on the "unread register" heuristic).

**Comment by InusualZ** [comment]
Thank you, seems to be working correctly as far as I can see.
